Specimen description and preparation step fixes

erikogabrielsson opened the pull request:
Hi,

I had some problems with using the 'SpecimenDescription` and `SpecimenPreparationStep` content items:
- The name for the DateTimeContentItem was incorrectly capitalized.
- Datetimes would not parse correctly.
- Some attributes in the [standard ](https://dicom.nema.org/medical/dicom/current/output/chtml/part03/sect_C.7.6.22.html ) were missing.

This PR fixes the above issues and adds tests for the bugs and added parameters. I'm not sureif the approach and style is in line with the rest of the code or if these changes breaks anything else. I'm happy to receive feedback if that is the case.

CPBridge commented:
Thanks @erikogabrielsson I'll take look!

CPBridge commented on tests/test_content.py:
```diff
@@ -625,6 +630,287 @@ def test_construction_staining(self):
+        assert (
+            instance.issuer_of_specimen_id ==
+            issuer_of_specimen_id.LocalNamespaceEntityID
+        )
```
Add check for type

erikogabrielsson replied:
Added a type check here. Here the type should be string, as per [TID 8001](https://dicom.nema.org/medical/dicom/current/output/chtml/part16/chapter_C.html#sect_TID_8001). Or should it be a `IssuerOfIdentifier`?

Also added a type check for the same-named property in `SpecimenDescription` (here it is a `IssuerOfIdentifier`). This might have been what you asked for, given [this](https://github.com/ImagingDataCommons/highdicom/pull/246#discussion_r1332386317) comment?

CPBridge replied:
Ah, yes I meant the second one. Sorry was getting the various similarly-named classes confused.

CPBridge commented on src/highdicom/sr/value_types.py:
```diff
@@ -877,19 +877,16 @@ def __init__(
+                raise ValueError(
+                    f'Could not decode time value "{self.Time}"'
+                ) from exception
+        return value.replace()
```
I didn't write this code originally (that was @hackermd ) and I'm not sure why it was written like it is nor why it doesn't work. Your approach seems more logical I have to say. But have you tried it with `pydicom.config.datetime_conversion`?

erikogabrielsson replied:
I added `config.datetime_conversion = datetime_conversion` as parameter to the `test_from_dataset` and it looks OK. Looking at the DA, TM, and DT inits they either take a string or a date, time, or datetime.

CPBridge reviewed (changes requested):
Hi @erikogabrielsson , sorry it took so long to look at this properly. You help with this is very much appreciated.

Overall looks great but I have some comments on the specifics

Thanks!

CPBridge commented:
Oh and one more request, sorry. Since this involves API changes could you please change the target branch of the pull request from `master` to `v0.22.0dev` and we'llaim to release the changes in the next minor release. Thanks

erikogabrielsson commented on src/highdicom/content.py:
```diff
@@ -1118,6 +1147,58 @@ def embedding_medium(self) -> Union[CodedConcept, None]:
+        return items[0].value
+
+    @property
+    def specimen_container(self) -> Union[CodedConcept, None]:
```
Added docstring similar to previous properties.

erikogabrielsson commented on src/highdicom/content.py:
```diff
@@ -1118,6 +1147,58 @@ def embedding_medium(self) -> Union[CodedConcept, None]:
+        return items[0].value
+
+    @property
+    def specimen_type(self) -> Union[CodedConcept, None]:
```
Added docstring similar to previous properties.

erikogabrielsson commented on src/highdicom/content.py:
```diff
@@ -1059,6 +1074,20 @@ def __init__(
             sequence.append(embedding_medium_item)
         self.SpecimenPreparationStepContentItemSequence = sequence
 
+        if specimen_container is not None:
+            specimen_container_item = CodeContentItem(
+                name=codes.SCT.SpecimenContainer,
+                value=specimen_container
+            )
+            sequence.append(specimen_container_item)
+
+        if specimen_type is not None:
+            specimen_type_item = CodeContentItem(
+                name=codes.SCT.SpecimenType,
+                value=specimen_type
+            )
+            sequence.append(specimen_type_item)
+
```
Yes, it it should definitely be before adding the sequence, together with the other steps.

erikogabrielsson commented on src/highdicom/content.py:
```diff
@@ -978,9 +981,21 @@ def __init__(
+        specimen_container: Union[pydicom.sr.coding.Code, highdicom.sr.CodedConcept], optional
+            Container the specimen resides in.
+        specimen_type: Union[pydicom.sr.coding.Code, highdicom.sr.CodedConcept], optional
+            The anatomic pathology specimen type of the specimen.
```
Added reference to CID for this and other arguments.

erikogabrielsson commented on src/highdicom/content.py:
```diff
@@ -1377,11 +1472,46 @@ def specimen_uid(self) -> UID:
         return UID(self.SpecimenUID)
 
+    @property
+    def specimen_location(self) -> Union[str, Tuple[float, float, float], None]:
```
Added docstring similar to previous properties.

erikogabrielsson commented on src/highdicom/content.py:
```diff
@@ -1377,11 +1472,46 @@ def specimen_uid(self) -> UID:
+        return self.get("SpecimenDetailedDescription")
+
+    @property
+    def issuer_of_specimen_id(self) -> Union[IssuerOfIdentifier, None]:
```
Added a  `from_dataset` method to the `IssuerOfIdentifier` class and used it in the `from_dataset` method of `SpecimenDescription`. Also added unit tests for construction.

erikogabrielsson commented on src/highdicom/content.py:
```diff
@@ -1268,6 +1349,9 @@ def __init__(
         ] = None,
+        specimen_type: Optional[Union[Code, CodedConcept]] = None,
+        specimen_short_description: Optional[str] = None,
+        specimen_detailed_description: Optional[str] = None,
```
Moved the new arguments to at the end of the argument lists.

erikogabrielsson commented on src/highdicom/content.py:
```diff
@@ -978,9 +981,21 @@ def __init__(
         embedding_medium: Union[pydicom.sr.coding.Code, highdicom.sr.CodedConcept], optional
             Embedding medium used during processing
+        specimen_container: Union[pydicom.sr.coding.Code, highdicom.sr.CodedConcept], optional
+            Container the specimen resides in.
```
I added CID references as suggested. Output looks ok after sphinx-build.

erikogabrielsson commented:
Hi @CPBridge and thanks for the review. No worries regarding the timing.

I think I have addressed all of your comments with new commits. Please re-review and let me know if anything was missed.

CPBridge commented on src/highdicom/content.py:
```diff
@@ -711,14 +712,72 @@ def __init__(
+        """highdicom.UniversalEntityIDTypeValues: Type of the issuer."""
+        return self._issuer_of_identifier_type
+
+    @classmethod
+    def from_dataset(
+        cls,
+        dataset: Dataset,
+    ) -> 'IssuerOfIdentifier':
+        """Construct object from an existing dataset.
+
+        Parameters
+        ----------
+        dataset: pydicom.dataset.Dataset
+            Dataset
+
+        Returns
+        -------
+        highdicom.IssuerOfIdentifier
+            Issuer of identifier
+
+        """
+        issuer_of_identifier = deepcopy(dataset)
```
For performance reasons, we have been adding the ability to bypass the copy into all `from_dataset` and `from_sequence` methods. I think weh should do the same here.

```suggestion
    @classmethod
    def from_dataset(
        cls,
        dataset: Dataset,
        copy: bool = True,
    ) -> 'IssuerOfIdentifier':
        """Construct object from an existing dataset.

        Parameters
        ----------
        dataset: pydicom.dataset.Dataset
            Dataset
        copy: bool
            If True, the underlying dataset is deep-copied such that the
            original dataset remains intact. If False, this operation will
            alter the original dataset in place.

        Returns
        -------
        highdicom.IssuerOfIdentifier
            Issuer of identifier

        """
        if copy:
            issuer_of_identifier = deepcopy(dataset)
        else:
            issuer_of_identifier = dataset
```


 (p.s. feel free to just accept the suggestion in github's interface).

CPBridge replied:
pps this won't be used furing the call from SpecimenDescription.from_dataset() yet, but I can add that later

erikogabrielsson replied:
Accepted and tested the suggestion.

CPBridge commented:
Thanks @erikogabrielsson, your changes look good. However it seems that some comments from my review may have been missed, probably because of github's annoying habit of "collapsing" comments in the middle of a review?

Specifically I'm talking about [this](https://github.com/ImagingDataCommons/highdicom/pull/246/files/87fc76053679df11f8e3a3b967834e6f5bfe6049#r1332375625), [this](https://github.com/ImagingDataCommons/highdicom/pull/246/files/87fc76053679df11f8e3a3b967834e6f5bfe6049#r1332379920) and [this](https://github.com/ImagingDataCommons/highdicom/pull/246/files/87fc76053679df11f8e3a3b967834e6f5bfe6049#r1332377113).

erikogabrielsson commented on src/highdicom/content.py:
```diff
@@ -1336,6 +1425,12 @@ def __init__(
+        if specimen_type is not None:
+            self.SpecimenTypeCodeSequence = [specimen_type]
+        if specimen_short_description is not None:
+            self.SpecimenShortDescription = specimen_short_description
```
Added suggestion and tests for invalid short descriptions (to long string, string with backslash).

erikogabrielsson commented on src/highdicom/content.py:
```diff
@@ -1289,6 +1373,12 @@ def __init__(
         specimen_preparation_steps: Sequence[highdicom.SpecimenPreparationStep], optional
             Steps that were applied during the preparation of the examined
             specimen in the laboratory prior to image acquisition
+        specimen_type: Union[pydicom.sr.coding.Code, highdicom.sr.CodedConcept], optional
+            The type of the examined specimen.
```
This one I missed, I changed to description on the `SpecimenPreparationStep`... :)

erikogabrielsson commented on src/highdicom/content.py:
```diff
@@ -1336,6 +1425,12 @@ def __init__(
             self.SpecimenLocalizationContentItemSequence = loc_seq
 
+        if specimen_type is not None:
+            self.SpecimenTypeCodeSequence = [specimen_type]
```
Added conversion of `Code` `specimen_type` to `CodedConcept` and test for constructing `SpecimenDescription` with  `Code` `specimen_type`.

erikogabrielsson commented:
Yes I had missed those (they were collapsed). I have now addressed the missed comments.

CPBridge reviewed (approved):
Thanks for your hard work on this @erikogabrielsson, we'll put these changes 0.22.0 (hopefullly not too long until I get around to this)

CPBridge commented:
PS I'll give @hackermd a few more days to comment if he likes before merging
